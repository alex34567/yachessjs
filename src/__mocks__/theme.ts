import DEFAULT_THEME from '../defaultTheme.json'
import assert from 'assert'
import { Theme } from '../theme'
import { useState } from 'react'

export class ThemeManager {
  readonly theme: Readonly<Theme>
  readonly className: string = 'TestTheme'

  readonly style = document.createElement('style')

  constructor (theme?: Theme) {
    if (!theme) {
      theme = DEFAULT_THEME
    }
    this.theme = { ...theme }
    assert(/^[0-9a-zA-Z#]+$/.test(this.theme.boardSquare.whiteColor), 'Attempted xss')
    assert(/^[0-9a-zA-Z#]+$/.test(this.theme.boardSquare.blackColor), 'Attempted xss')
    this.style.textContent = `
      .ChessBoardSquareWhite.${this.className} {
        background-color: ${this.theme.boardSquare.whiteColor};
      }

      .ChessBoardSquareBlack.${this.className} {
       background-color: ${this.theme.boardSquare.blackColor};
      }
    `
  }
}

export function useTheme (defaultTheme?: Theme): [ThemeManager, (theme: Theme) => void] {
  const [theme, setTheme] = useState(new ThemeManager(defaultTheme))

  return [theme, theme => {
    setTheme(new ThemeManager(theme))
  }]
}
