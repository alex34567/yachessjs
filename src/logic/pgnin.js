'use strict'

const pieces = require('./pieces')
const state = require('./state')
const moveNotation = require('./move_notation')

const PGN_TOKEN_REGEX = /([[\].*()<>]|"(?:[^\\"]|\\"|\\\\)+"|$\d+|[A-Za-z0-9][A-Za-z0-9_+#=:\-!?/]*)|{.*?}|;.*$|^%.*|\s+/my
const TAGNAME_REGEX = /[a-zA-Z0-9_]+/

class PGNGame {
  constructor (tags, state) {
    this.tags = tags
    this.state = state
  }
}

function parsePgn (pgnStr) {
  const db = []
  const tokenStream = pgnStr.split(new RegExp(PGN_TOKEN_REGEX)).filter(x => x)
  let tokenIndex = 0

  while (tokenIndex < tokenStream.length) {
    const tags = []
    while (tokenStream[tokenIndex] === '[') {
      tokenIndex++
      const name = tokenStream[tokenIndex]
      tokenIndex++
      if (!TAGNAME_REGEX.test(name)) {
        throw new Error('PGN Parse Error')
      }
      const rawValue = tokenStream[tokenIndex]
      tokenIndex++
      if (rawValue[0] !== '"') {
        throw new Error('PGN never ending string')
      }
      const value = rawValue.substring(1, rawValue.length - 1)
      tags.push([name, value])
      if (tokenStream[tokenIndex] !== ']') {
        throw new Error('PGN Parse Error')
      }
      tokenIndex++
    }
    const setupIndex = tags.findIndex(x => x[0] === 'SetUp')
    const fenIndex = tags.findIndex(x => x[0] === 'FEN')
    let gameState = state.START_STATE
    if (setupIndex >= 0 && tags[setupIndex][1] === '1' && fenIndex >= 0) {
      gameState = state.stateFromFen(tags[fenIndex][1])
    }
    while (true) {
      tokenIndex++
      if (/^[0-9]+$/.test(tokenStream[tokenIndex - 1])) {
        if (String(gameState.moveCount()) !== tokenStream[tokenIndex - 1]) {
          throw new Error('Invalid game number enumeration')
        }
        continue
      }
      if (tokenStream[tokenIndex - 1][0] === '$') {
        continue
      }
      let done = false
      switch (tokenStream[tokenIndex - 1]) {
        case '(': {
          let depth = 1
          while (depth > 0) {
            if (tokenStream[tokenIndex] === '(') {
              depth++
            } else if (tokenStream[tokenIndex] === ')') {
              depth--
            }
            tokenIndex++
          }
          continue
        }
        case '.':
          continue
        case '1-0':
          if (gameState.currTurn !== pieces.BLACK) {
            throw new Error('Invalid result')
          }
          if (!gameState.isCheckmate()) {
            gameState = gameState.modify(x => { x.resign = true })
          }
          done = true
          break
        case '0-1':
          if (gameState.currTurn !== pieces.WHITE) {
            throw new Error('Invalid result')
          }
          if (!gameState.isCheckmate()) {
            gameState = gameState.modify(x => { x.resign = true })
          }
          done = true
          break
        case '1/2-1/2':
          if (!gameState.isDraw()) {
            gameState = gameState.modify(x => { x.agreeDraw = true })
          }
          done = true
          break
        case '*':
          done = true
          break
      }
      if (done) {
        break
      }
      const move = tokenStream[tokenIndex - 1]
      const parsedMove = moveNotation.MoveNotation.parseMove(gameState.currTurn, move)
      if (!parsedMove) {
        throw new Error('Move syntax error')
      }
      const possibleMoves = gameState.moves().filter(x => !x.invalid() && parsedMove.matches(x))
      if (possibleMoves.length === 0) {
        throw new Error('Invalid move tried to be played')
      }
      if (possibleMoves.length > 1) {
        throw new Error('Move ambiguity in PGN file')
      }
      gameState = possibleMoves[0].do()
    }
    db.push(new PGNGame(tags, gameState))
  }

  return db
}

Object.assign(module.exports, { parsePgn })
