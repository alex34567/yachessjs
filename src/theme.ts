import { useEffect, useState } from 'react'
import DEFAULT_THEME from './defaultTheme.json'
import assert from 'assert'

export interface BoardSquareTheme {
  name: string,
  id: string,
  // Must be an css color
  blackColor: string,
  whiteColor: string,
}

export interface PieceTheme {
  prefix: string,
  name: string,
}

export interface Theme {
  piece: PieceTheme
  boardSquare: BoardSquareTheme,
}

let CLASSNAME_SUFFIX = 0

export class ThemeManager {
  readonly theme: Readonly<Theme>
  readonly className: string

  readonly style = document.createElement('style')

  constructor (theme?: Theme) {
    if (!theme) {
      theme = DEFAULT_THEME
    }
    this.theme = { ...theme }
    this.className = `ChessTheme-${CLASSNAME_SUFFIX}`
    CLASSNAME_SUFFIX++
    assert(/^[0-9a-zA-Z#]+$/.test(this.theme.boardSquare.whiteColor), 'Attempted xss')
    assert(/^[0-9a-zA-Z#]+$/.test(this.theme.boardSquare.blackColor), 'Attempted xss')
    this.style.textContent = `
      .ChessBoardSquareWhite.${this.className} {
        background-color: ${this.theme.boardSquare.whiteColor};
      }

      .ChessBoardSquareBlack.${this.className} {
       background-color: ${this.theme.boardSquare.blackColor};
      }
    `
  }
}

export function useTheme (defaultTheme?: Theme): [ThemeManager, (theme: Theme) => void] {
  const [theme, setTheme] = useState(new ThemeManager(defaultTheme))
  useEffect(() => {
    document.head.append(theme.style)
    return () => {
      theme.style.remove()
    }
  }, [theme])

  return [theme, theme => {
    setTheme(new ThemeManager(theme))
  }]
}
