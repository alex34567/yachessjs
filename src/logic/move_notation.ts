import * as util from './util'
import * as moves from './moves'
import * as pieces from './pieces'

const NORMAL_MOVE_REGEX = /^([KQRBN]?)([a-h]?)([1-8]?)(x?)([a-h])([1-8])(?:=?([QRBN]))?(?:\+|#|)(?:\?\??|!!?|\?!|!\?|)$/

abstract class MoveNotation {
  protected constructor () {
    if (typeof this.matches !== 'function') {
      throw new Error('Matches function not defined')
    }
  }

  static parseMove (color: pieces.Color, move: string) {
    move = move.trim()
    if (move === 'O-O' || move === '0-0') {
      return new CastleMoveNotation(true)
    }
    if (move === 'O-O-O' || move === '0-0-0') {
      return new CastleMoveNotation(false)
    }
    const parsedMove = NORMAL_MOVE_REGEX.exec(move)
    if (!parsedMove) {
      return null
    }

    const pieceLetter = parsedMove[1]
    const piece = color.FROM_LETTER.get(pieceLetter)!
    const fromFileLetter = parsedMove[2]
    const fromFile = fromFileLetter ? util.convertFileLetter(fromFileLetter) : null
    const fromRankLetter = parsedMove[3]
    const fromRank = fromRankLetter ? util.convertRankLetter(fromRankLetter) : null
    const captures = parsedMove[4] === 'x'
    const toFile = util.convertFileLetter(parsedMove[5])
    const toRank = util.convertRankLetter(parsedMove[6])
    const toPos = new util.Pos(toFile, toRank)
    const promoteLetter = parsedMove[7]
    let promoteChoice = null
    if (promoteLetter) {
      promoteChoice = color.FROM_LETTER.get(promoteLetter)!
    }
    return new NormalMoveNotation(piece, captures, fromFile, fromRank, toPos, promoteChoice)
  }

  abstract matches (move: moves.Move): boolean
}

class CastleMoveNotation extends MoveNotation {
  isKingSide: boolean

  constructor (isKingSide: boolean) {
    super()
    this.isKingSide = isKingSide
  }

  toString () {
    if (this.isKingSide) {
      return 'O-O'
    } else {
      return 'O-O-O'
    }
  }

  matches (move: moves.Move): boolean {
    return move.isCastle() && this.isKingSide === move.isKingSide
  }
}

class NormalMoveNotation extends MoveNotation {
  piece: pieces.Piece
  captures: boolean
  fromFile: number | null
  fromRank: number | null
  toPos: util.Pos
  promoteChoice: pieces.Piece | null
  checkSymbol: string

  constructor (piece: pieces.Piece, captures: boolean, fromFile: number | null, fromRank: number | null, toPos: util.Pos, promoteChoice: pieces.Piece | null) {
    super()
    this.piece = piece
    this.captures = captures
    this.fromFile = fromFile
    this.fromRank = fromRank
    this.toPos = toPos
    this.promoteChoice = promoteChoice
    this.checkSymbol = ''
  }

  toString () {
    let captureLetter = ''
    if (this.captures) {
      captureLetter = 'x'
    }
    let fromFileLetter = ''
    if (this.fromFile !== null) {
      fromFileLetter = String.fromCharCode(this.fromFile + 'a'.charCodeAt(0))
    }
    let fromRankDigit = ''
    if (this.fromRank !== null) {
      fromRankDigit = String(this.fromRank + 1)
    }
    let promoteStr = ''
    if (this.promoteChoice) {
      promoteStr += '=' + this.promoteChoice.getPGNLetter()
    }
    return this.piece.getPGNLetter() + fromFileLetter + fromRankDigit + captureLetter + this.toPos.toString() + promoteStr + this.checkSymbol
  }

  matches (move: moves.Move): boolean {
    // Promotes are only valid for pawns
    if (this.piece !== move.state.currTurn.PAWN && this.promoteChoice) {
      return false
    }
    if (!move.isNormal()) {
      return false
    }
    // Be permissive in allowing a capture when not specified but require that a move like Qxe6 is a capture
    if (this.captures && !(move.state.board.get(this.toPos).canBeCaptured(move.piece) || move.isEnPassant())) {
      return false
    }
    if (move.isPromote() && this.promoteChoice !== move.promoteChoice) {
      return false
    }
    if (!move.isPromote() && this.promoteChoice) {
      return false
    }
    return move.piece === this.piece &&
            (this.fromFile === null || this.fromFile === move.fromPos.file) &&
            (this.fromRank === null || this.fromRank === move.fromPos.rank) &&
            move.toPos.compare(this.toPos) === 0
  }
}

export { CastleMoveNotation, MoveNotation, NormalMoveNotation }
