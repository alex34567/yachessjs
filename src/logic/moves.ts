import * as pieces from './pieces'
import * as util from './util'
import * as moveNotation from './move_notation'
import * as state from './state'
import * as immmutable from 'immutable'
import { State } from './state'
import assert from 'assert'

abstract class Move {
  state: state.State

  protected constructor (state: state.State) {
    this.state = state
    if (typeof this.toNotation !== 'function') {
      throw new Error('To notation function not defined')
    }
  }

  isNormal (): this is NormalMove {
    return false
  }

  isCastle (): this is Castle {
    return false
  }

  isEnPassant (): this is EnPassant {
    return false
  }

  do () {
    return this.state.flipTurn().modify(newState => this.doChain(newState))
  }

  protected doChain (state: State) {
    state.halfMove++
    state.moveHistory = state.moveHistory.push(this)
    const lastFen = this.state.toAbvFen()
    let stateCount = state.threeFoldDetect.get(lastFen)
    if (!stateCount) {
      stateCount = 0
    }
    stateCount++
    state.threeFoldDetect = state.threeFoldDetect.set(lastFen, stateCount)
  }

  toString () {
    return this.toNotation().toString()
  }

  abstract invalid (): string | false

  abstract toNotation(): moveNotation.MoveNotation
}

class NormalMove extends Move {
  piece: pieces.Piece
  fromPos: util.Pos
  toPos: util.Pos
  private notationCache?: moveNotation.MoveNotation

  constructor (state: state.State, piece: pieces.Piece, fromPos: util.Pos, toPos: util.Pos) {
    super(state)
    this.piece = piece
    this.fromPos = fromPos
    this.toPos = toPos
  }

  isNormal () {
    return true
  }

  isPromote (): this is Promotion {
    return false
  }

  isCapture () {
    return this.state.board.get(this.toPos).isOccupied()
  }

  toNotation () {
    if (this.notationCache) {
      return this.notationCache
    }
    let promoteChoice = null
    if (this.isPromote()) {
      promoteChoice = this.promoteChoice
    }
    const notation = new moveNotation.NormalMoveNotation(this.piece, this.isCapture(), null, null, this.toPos, promoteChoice)
    let possibleMoves = this.state.moves().filter(x => notation.matches(x) && !x.invalid())
    if (possibleMoves.length > 1 || (this.isCapture() && this.piece === this.piece.color.PAWN)) {
      notation.fromFile = this.fromPos.file
      possibleMoves = this.state.moves().filter(x => notation.matches(x) && !x.invalid())
      if (possibleMoves.length > 1) {
        notation.fromFile = null
        notation.fromRank = this.fromPos.rank
        possibleMoves = this.state.moves().filter(x => notation.matches(x) && !x.invalid())
        if (possibleMoves.length > 1) {
          notation.fromFile = this.fromPos.file
          notation.fromRank = this.fromPos.rank
        }
      }
    }
    const doneMove = this.do()
    if (doneMove.isCheckmate()) {
      notation.checkSymbol = '#'
    } else if (doneMove.isCheck()) {
      notation.checkSymbol = '+'
    }
    this.notationCache = notation
    return notation
  }

  invalid (): string | false {
    if (this.piece === this.state.currTurn.KING) {
      if (this.state.board.isProtected(this.toPos)) {
        return 'Check'
      }
      return false
    } else {
      // In double check only the king is allowed to move
      if (this.state.checks > 1) {
        return 'Check'
      }
      if (this.state.board.isPinned(this.fromPos)) {
        const deltaFile = this.fromPos.file - this.toPos.file
        const deltaRank = this.fromPos.rank - this.toPos.rank
        const knightMove = deltaFile !== 0 && deltaRank !== 0 && Math.abs(deltaFile) !== Math.abs(deltaRank)
        if (knightMove) {
          return 'Check'
        }
        const pinnedAxis = this.state.board.getPinnedAxis(this.fromPos)
        if (deltaFile === 0 && pinnedAxis !== 0) {
          return 'Check'
        }
        if (deltaRank === 0 && pinnedAxis !== 1) {
          return 'Check'
        }
        if (deltaFile !== 0 && deltaRank !== 0) {
          if (Math.sign(deltaRank) !== Math.sign(deltaFile) && pinnedAxis !== 2) {
            return 'Check'
          }
          if (Math.sign(deltaRank) === Math.sign(deltaFile) && pinnedAxis !== 3) {
            return 'Check'
          }
        }
      }

      // Piece must block check or capture checking piece
      if (this.state.checks === 1 && !this.state.board.isPinned(this.toPos)) {
        return 'Check'
      }
      return false
    }
  }

  protected doChain (state: State) {
    super.doChain(state)
    state.board = state.board.set(this.fromPos, pieces.EMPTY)
    state.board = state.board.set(this.toPos, this.piece)
    if (this.state.board.get(this.toPos).isOccupied() || this.piece === this.state.currTurn.PAWN) {
      state.halfMove = 0
      state.threeFoldDetect = immmutable.Map()
    }
  }
}

class FirstPawn extends NormalMove {
  protected doChain (state: State) {
    super.doChain(state)
    const myColor = state.getColor(this.piece.color)
    myColor.enPassantPos = this.fromPos.addRank(this.piece.color.PAWN_RANK_DIR)
  }
}

class EnPassant extends NormalMove {
  isEnPassant () {
    return true
  }

  isCapture () {
    return true
  }

  protected doChain (state: State) {
    super.doChain(state)
    const passedPawn = this.toPos.addRank(-this.piece.color.PAWN_RANK_DIR)
    assert(passedPawn)

    state.board = state.board.set(passedPawn, pieces.EMPTY)
  }
}

class Promotion extends NormalMove {
  promoteChoice: pieces.Piece

  constructor (state: state.State, piece: pieces.Piece, fromPos: util.Pos, toPos: util.Pos, promoteChoice: pieces.Piece) {
    super(state, piece, fromPos, toPos)
    this.promoteChoice = promoteChoice
  }

  isPromote () {
    return true
  }

  doChain (state: State) {
    super.doChain(state)
    state.board = state.board.set(this.toPos, this.promoteChoice)
  }
}

class Castle extends Move {
  isKingSide: boolean

  constructor (state: state.State, isKingSide: boolean) {
    super(state)
    this.isKingSide = isKingSide
  }

  toNotation () {
    const notation = new moveNotation.CastleMoveNotation(this.isKingSide)
    const doneMove = this.do()
    if (doneMove.isCheckmate()) {
      notation.checkSymbol = '#'
    } else if (doneMove.isCheck()) {
      notation.checkSymbol = '+'
    }
    return notation
  }

  isCastle () {
    return true
  }

  doChain (state: State) {
    super.doChain(state)
    const myRook = this.state.currTurn.ROOK
    const myKing = this.state.currTurn.KING
    const myRank = this.state.currTurn.KING_RANK

    state.board = state.board.withMutations(board => {
      board.set(new util.Pos(4, myRank), pieces.EMPTY)
      if (this.isKingSide) {
        board.set(new util.Pos(5, myRank), myRook)
        board.set(new util.Pos(6, myRank), myKing)
        board.set(new util.Pos(7, myRank), pieces.EMPTY)
      } else {
        board.set(new util.Pos(3, myRank), myRook)
        board.set(new util.Pos(2, myRank), myKing)
        board.set(new util.Pos(0, myRank), pieces.EMPTY)
      }
    })
  }

  invalid () {
    if (this.state.isCheck()) {
      return "Can't castle out of check"
    }

    const myColor = this.state.getColor(this.state.currTurn)
    const myRank = this.state.currTurn.KING_RANK

    const betweenRookKing = []
    if (this.isKingSide) {
      if (!myColor.kingSideCastle) {
        return "Can't castle on king side"
      }

      betweenRookKing.push(new util.Pos(5, myRank), new util.Pos(6, myRank))
    } else {
      if (!myColor.queenSideCastle) {
        return "Can't castle on queen side"
      }

      betweenRookKing.push(new util.Pos(1, myRank), new util.Pos(2, myRank), new util.Pos(3, myRank))
    }

    for (const pos of betweenRookKing) {
      if (this.state.board.get(pos).isOccupied()) {
        return 'Pieces are between rook and king'
      }
    }

    const badPos: util.Pos[] = []
    if (this.isKingSide) {
      // Can't castle though check
      badPos.push(new util.Pos(5, myRank))
      // Can't castle into check
      badPos.push(new util.Pos(6, myRank))
    } else {
      // Can't castle into check
      badPos.push(new util.Pos(2, myRank))
      // Can't castle though check
      badPos.push(new util.Pos(3, myRank))
    }

    const attacked = badPos.some(pos => this.state.board.isProtected(pos))
    if (attacked) {
      return "Can't castle into or through check"
    }
    return false
  }
}

export { Castle, Move, EnPassant, NormalMove, FirstPawn, Promotion }
