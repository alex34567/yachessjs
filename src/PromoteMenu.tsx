import * as immutable from 'immutable'
import { Promotion } from './logic/moves'
import { State } from './logic/state'
import BoardSquare from './BoardSquare'
import * as React from 'react'

interface PromoteMenuProps {
  moves: immutable.List<Promotion>,
  onPromote: (state: State) => void,
}

export default function PromoteMenu (props: PromoteMenuProps) {
  const promotes = []
  let key = 0
  for (const move of props.moves) {
    const onClick = () => {
      props.onPromote(move.do())
    }
    promotes.push(<BoardSquare key={key++} piece={move.promoteChoice} onClick={onClick} />)
  }
  return (
    <div className='PromoteMenu'>
      {promotes}
    </div>
  )
}
