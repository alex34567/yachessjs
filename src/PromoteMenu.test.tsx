import { render, unmountComponentAtNode } from 'react-dom'
import { stateFromFen } from './logic/state'
import { Promotion } from './logic/moves'
import { act } from '@testing-library/react'
import PromoteMenu from './PromoteMenu'
import * as immutable from 'immutable'
import React from 'react'
import { ThemeManager } from './theme'

jest.mock('./BoardSquare')
jest.mock('./theme')

let container: HTMLDivElement | null = null
beforeEach(() => {
  // setup a DOM element as a render target
  container = document.createElement('div')
  document.body.appendChild(container)
})

afterEach(() => {
  // cleanup on exiting
  unmountComponentAtNode(container!)
  container!.remove()
  container = null
})

test('render white', () => {
  const state = stateFromFen('8/P6k/8/8/8/8/8/7K w - - 0 1')
  const moves = immutable.List(state.moves().filter(x => !x.invalid() && x.isNormal() && x.isPromote()) as Promotion[])
  const onPromote = jest.fn()
  act(() => {
    render(<PromoteMenu moves={moves} onPromote={onPromote} theme={new ThemeManager()} />, container)
  })
  expect(container).toMatchSnapshot()
})

test('render black', () => {
  const state = stateFromFen('8/7k/8/8/8/8/p7/7K b - - 0 1')
  const moves = immutable.List(state.moves().filter(x => !x.invalid() && x.isNormal() && x.isPromote()) as Promotion[])
  const onPromote = jest.fn()
  act(() => {
    render(<PromoteMenu moves={moves} onPromote={onPromote} theme={new ThemeManager()} />, container)
  })
  expect(container).toMatchSnapshot()
})

test('queen promote', () => {
  const state = stateFromFen('8/P6k/8/8/8/8/8/7K w - - 0 1')
  const moves = immutable.List(state.moves().filter(x => !x.invalid() && x.isNormal() && x.isPromote()) as Promotion[])
  const onPromote = jest.fn()
  act(() => {
    render(<PromoteMenu moves={moves} onPromote={onPromote} theme={new ThemeManager()} />, container)
  })

  act(() => {
    container?.firstElementChild?.children[0].dispatchEvent(new MouseEvent('click', { bubbles: true }))
  })

  expect(onPromote.mock.calls.length).toBe(1)
  expect(onPromote.mock.calls[0][0].toFen()).toBe('Q7/7k/8/8/8/8/8/7K b - - 0 1')
})

test('knight promote', () => {
  const state = stateFromFen('8/P6k/8/8/8/8/8/7K w - - 0 1')
  const moves = immutable.List(state.moves().filter(x => !x.invalid() && x.isNormal() && x.isPromote()) as Promotion[])
  const onPromote = jest.fn()
  act(() => {
    render(<PromoteMenu moves={moves} onPromote={onPromote} theme={new ThemeManager()} />, container)
  })

  act(() => {
    container?.firstElementChild?.children[1].dispatchEvent(new MouseEvent('click', { bubbles: true }))
  })

  expect(onPromote.mock.calls.length).toBe(1)
  expect(onPromote.mock.calls[0][0].toFen()).toBe('N7/7k/8/8/8/8/8/7K b - - 0 1')
})

test('rook promote', () => {
  const state = stateFromFen('8/P6k/8/8/8/8/8/7K w - - 0 1')
  const moves = immutable.List(state.moves().filter(x => !x.invalid() && x.isNormal() && x.isPromote()) as Promotion[])
  const onPromote = jest.fn()
  act(() => {
    render(<PromoteMenu moves={moves} onPromote={onPromote} theme={new ThemeManager()} />, container)
  })

  act(() => {
    container?.firstElementChild?.children[2].dispatchEvent(new MouseEvent('click', { bubbles: true }))
  })

  expect(onPromote.mock.calls.length).toBe(1)
  expect(onPromote.mock.calls[0][0].toFen()).toBe('R7/7k/8/8/8/8/8/7K b - - 0 1')
})

test('bishop promote', () => {
  const state = stateFromFen('8/P6k/8/8/8/8/8/7K w - - 0 1')
  const moves = immutable.List(state.moves().filter(x => !x.invalid() && x.isNormal() && x.isPromote()) as Promotion[])
  const onPromote = jest.fn()
  act(() => {
    render(<PromoteMenu moves={moves} onPromote={onPromote} theme={new ThemeManager()} />, container)
  })

  act(() => {
    container?.firstElementChild?.children[3].dispatchEvent(new MouseEvent('click', { bubbles: true }))
  })

  expect(onPromote.mock.calls.length).toBe(1)
  expect(onPromote.mock.calls[0][0].toFen()).toBe('B7/7k/8/8/8/8/8/7K b - - 0 1')
})
