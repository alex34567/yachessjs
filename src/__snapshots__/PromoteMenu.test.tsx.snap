// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render black 1`] = `
<div>
  <div
    class="PromoteMenu"
  >
    <div
      data-name="q"
    />
    <div
      data-name="n"
    />
    <div
      data-name="r"
    />
    <div
      data-name="b"
    />
  </div>
</div>
`;

exports[`render white 1`] = `
<div>
  <div
    class="PromoteMenu"
  >
    <div
      data-name="Q"
    />
    <div
      data-name="N"
    />
    <div
      data-name="R"
    />
    <div
      data-name="B"
    />
  </div>
</div>
`;
